R/utilities.R

```diff
@@ -1148,3 +1148,185 @@ ToNumeric <- function(x){
+  }
+
+  order <- Cells(object = object)
+  enum.fields <- c()
+
+  # Export expression matrix
+  df <- as.data.frame(as.matrix(GetAssayData(object = object)))
+  df <- data.frame(gene=rownames(object), df)
+  z <- gzfile(file.path(dir, "exprMatrix.tsv.gz"), "w")
+  write.table(df, sep="\t", file=z, quote = FALSE, row.names = FALSE)
+  close(z)
+
+  # Export cell embeddings
+  embeddings.conf <- c()
+  for (embedding in embeddings) {
+    df <- Embeddings(object = object, reduction = embedding)
+    if (ncol(df) > 2) {
+      warning(sprintf(
+        'Embedding %s has more than 2 coordinates, taking only the first 2',
+        embedding
+      ))
+      df <- df[, 1:2]
+    }
+    colnames(df) <- c("x", "y")
+    df <- data.frame(cellId = rownames(df), df)
+    fname <- file.path(
+      dir,
+      sprintf("%s.coords.tsv", embedding)
+    )
+    write.table(df[order, ], sep="\t", file=fname, quote = FALSE, row.names = FALSE)
+    conf <- sprintf(
+      '{"file": "%s.coords.tsv", "shortLabel": "Seurat %1$s"}',
+      embedding
+    )
+    embeddings.conf <- c(embeddings.conf, conf)
+  }
+
+  # Export metadata
+  df <- data.frame(row.names = rownames(object[[]]))
+  for (field in meta.fields) {
+    if (field == ".ident") {
+      df$Cluster <- Idents(object = object)
+      enum.fields <- c(enum.fields, "Cluster")
+      cluster.field <- "Cluster"
+    } else {
+      name <- meta.fields.names[[field]]
+      if (is.null(name)) {
+        name <- field
+      }
+      df[[name]] <- object[[field]][, 1]
+      if (!is.numeric(df[[name]])) {
+        enum.fields <- c(enum.fields, name)
+      }
+    }
+  }
+  df <- data.frame(Cell=rownames(df), df)
+  write.table(df[order, ], sep="\t", file=file.path(dir, "meta.tsv"), quote = FALSE, row.names=FALSE)
+
+  # Export markers
+  markers.string <- ''
+  if (!is.null(markers.file)) {
+    file.copy(markers.file, file.path(dir, "markers.tsv"))
+    markers.string <- 'markers = [{"file": "markers.tsv", "shortLabel": "Seurat Cluster Markers"}]'
+  }
+
+  config <- 'name="%s"
+shortLabel="%1$s"
+exprMatrix="exprMatrix.tsv.gz"
+#tags = ["10x", "smartseq2"]
+meta="meta.tsv"
+# possible values: "gencode-human", "gencode-mouse", "symbol" or "auto"
+geneIdType="auto"
+clusterField="%s"
+labelField="%2$s"
+enumFields=%s
+%s
+coords=%s'
+  enum.string <- paste0(
+    "[",
+    paste(paste0('"', enum.fields, '"'), collapse = ", "),
+    "]"
+  )
+  coords.string <- paste0(
+    "[",
+    paste(embeddings.conf, collapse = ",\n"),
+    "]"
+  )
+  config <- sprintf(
+    config,
+    dataset.name,
+    cluster.field,
+    enum.string,
+    markers.string,
+    coords.string
+  )
+  cat(config, file=file.path(dir, "cellbrowser.conf"))
+  message("Prepared cellbrowser directory ", dir)
+
+  if (!is.null(cb.dir) && require(reticulate)) {
+    library(reticulate)
+    if (!py_module_available("cellbrowser")) {
+      py_install("cellbrowser") # see https://rstudio.github.io/reticulate/articles/python_packages.html
```

Comment:
If the module isn't available, we would prefer that the function stops with a message rather than trying to install with `py_install`. Users tend to have varied preferred installation methods and would prefer that they handle the install themselves.

Reply:
Thanks for the detailed review!

I agree with you on this on, but there is one problem that is solved by installing the package through reticulate: users may have multiple pythons (2, 3, the one that came with the OS, one they built themselves, one through conda, etc) and they may not be sure which one is the python that is used by R. So by installing the package through reticulate, we are sure that we're installing into the correct Python. I agree that it shouldn't do the install automatically, but maybe we can add another parameter do.install=FALSE that the user can set to get the installation done? If the package is not there, we can stop with a message that tells the user to use pip or alternative set do.install=TRUE? Would you be OK with that?